File: apps/runtime/nodejs/sdk-compat-check.mjs
// SDK version alignment check — compares the app's declared @celerity-sdk/*
// dependency ranges against the runtime's installed versions.
// Warns on mismatch but never blocks startup.

import { readFileSync, readdirSync } from "node:fs";
import { join, resolve } from "node:path";

const LOG_PREFIX = "[celerity-runtime]";

/**
 * Simple semver range check for ^x.y.z and ~x.y.z patterns.
 * Compares major.minor without requiring a semver dependency.
 */
function satisfiesRange(installed, range) {
  const baseMatch = range.match(/(\d+)\.(\d+)\.(\d+)/);
  const instMatch = installed.match(/(\d+)\.(\d+)\.(\d+)/);
  if (!baseMatch || !instMatch) return true; // Can't parse, assume OK

  const baseMajor = Number(baseMatch[1]);
  const baseMinor = Number(baseMatch[2]);
  const basePatch = Number(baseMatch[3]);
  const instMajor = Number(instMatch[1]);
  const instMinor = Number(instMatch[2]);
  const instPatch = Number(instMatch[3]);

  if (range.startsWith("^")) {
    // ^x.y.z: same major, installed >= base
    return (
      instMajor === baseMajor &&
      (instMinor > baseMinor || (instMinor === baseMinor && instPatch >= basePatch))
    );
  }

  if (range.startsWith("~")) {
    // ~x.y.z: same major.minor, installed patch >= base patch
    return instMajor === baseMajor && instMinor === baseMinor && instPatch >= basePatch;
  }

  // Exact or other — just check major.minor match
  return instMajor === baseMajor && instMinor === baseMinor;
}

/**
 * Locate the app's package.json by walking up from CELERITY_MODULE_PATH,
 * falling back to CELERITY_APP_DIR.
 */
function findAppPackageJson() {
  const appDir = process.env.CELERITY_APP_DIR || "./app";
  const modulePath = process.env.CELERITY_MODULE_PATH;

  const searchDirs = [];
  if (modulePath) {
    let dir = join(resolve(modulePath), "..");
    for (let i = 0; i < 5; i++) {
      searchDirs.push(dir);
      dir = join(dir, "..");
    }
  }
  searchDirs.push(resolve(appDir));

  for (const dir of searchDirs) {
    try {
      return JSON.parse(readFileSync(join(dir, "package.json"), "utf8"));
    } catch {
      // Not found in this directory, try next
    }
  }
  return null;
}

/**
 * Read installed @celerity-sdk/* versions from the runtime's node_modules.
 * Returns a map of package short name to version string.
 *
 * @param {URL | string} runtimeBaseUrl - import.meta.url of the calling module
 */
export function getRuntimeSdkVersions(runtimeBaseUrl) {
  const sdkDir = new URL("./node_modules/@celerity-sdk", runtimeBaseUrl).pathname;
  const versions = {};
  try {
    for (const pkg of readdirSync(sdkDir)) {
      try {
        const pkgJson = JSON.parse(
          readFileSync(join(sdkDir, pkg, "package.json"), "utf8"),
        );
        versions[pkg] = pkgJson.version;
      } catch {
        // Skip platform-specific binaries without a standard package.json
      }
    }
  } catch {
    // @celerity-sdk directory not found
  }
  return versions;
}

/**
 * Compare the app's declared @celerity-sdk/* dependency ranges against
 * the runtime's installed versions. Logs warnings on mismatch.
 *
 * @param {URL | string} runtimeBaseUrl - import.meta.url of the calling module
 */
export function checkSdkVersionAlignment(runtimeBaseUrl) {
  const appPkg = findAppPackageJson();
  if (!appPkg) return;

  const appDeps = {
    ...appPkg.dependencies,
    ...appPkg.peerDependencies,
    ...appPkg.devDependencies,
  };

  const runtimeNodeModules = new URL("./node_modules", runtimeBaseUrl).pathname;
  const mismatches = [];

  for (const [name, range] of Object.entries(appDeps)) {
    if (!name.startsWith("@celerity-sdk/") || typeof range !== "string") continue;

    try {
      const installedPkg = JSON.parse(
        readFileSync(join(runtimeNodeModules, name, "package.json"), "utf8"),
      );
      if (!satisfiesRange(installedPkg.version, range)) {
        mismatches.push({ name, declared: range, installed: installedPkg.version });
      }
    } catch {
      // Package not in runtime's node_modules (e.g., @celerity-sdk/bucket) — skip
    }
  }

  if (mismatches.length > 0) {
    console.warn(`${LOG_PREFIX} SDK version mismatch detected:`);
    for (const m of mismatches) {
      console.warn(`  ${m.name}: app declares ${m.declared}, runtime provides ${m.installed}`);
    }
    console.warn(
      `${LOG_PREFIX} Consider aligning your app's SDK versions with the runtime.`,
    );
  }
}

File: apps/runtime/python/sdk_compat_check.py
# SDK version alignment check — compares the app's declared celerity-sdk
# dependency ranges against the runtime's installed versions.
# Warns on mismatch but never blocks startup.

from __future__ import annotations

import contextlib
import os
import re
import tomllib
from importlib.metadata import PackageNotFoundError, version
from pathlib import Path
from typing import Any

LOG_PREFIX = "[celerity-runtime]"

SDK_PACKAGES = ["celerity-sdk", "celerity-runtime-sdk"]


def get_runtime_sdk_versions() -> dict[str, str]:
    """Return installed versions of celerity SDK packages."""
    versions: dict[str, str] = {}
    for pkg in SDK_PACKAGES:
        with contextlib.suppress(PackageNotFoundError):
            versions[pkg] = version(pkg)
    return versions


def _find_app_pyproject() -> dict[str, Any] | None:
    """Locate the app's pyproject.toml by searching from CELERITY_MODULE_PATH."""

    app_dir = os.environ.get("CELERITY_APP_DIR", "./app")
    module_path = os.environ.get("CELERITY_MODULE_PATH")

    search_dirs: list[Path] = []
    if module_path:
        d = Path(module_path).resolve().parent
        for _ in range(5):
            search_dirs.append(d)
            d = d.parent
    search_dirs.append(Path(app_dir).resolve())

    for d in search_dirs:
        pyproject = d / "pyproject.toml"
        if pyproject.is_file():
            with open(pyproject, "rb") as f:
                return tomllib.load(f)
    return None


def _parse_version_from_spec(spec: str) -> tuple[str, str] | None:
    """Extract operator and version from a PEP 440 specifier like '>=0.2.0'."""
    match = re.match(r"([><=!~]+)\s*(\d+\.\d+\.\d+)", spec)
    if match:
        return match.group(1), match.group(2)
    return None


def _satisfies_range(installed: str, spec: str) -> bool:
    """Simple version range check for >=x.y.z patterns."""
    parsed = _parse_version_from_spec(spec)
    if not parsed:
        return True  # Can't parse, assume OK

    op, base = parsed
    inst_parts = [int(x) for x in installed.split(".")[:3]]
    base_parts = [int(x) for x in base.split(".")[:3]]

    if op == ">=":
        return inst_parts >= base_parts
    if op in ("==", "==="):
        return inst_parts == base_parts
    if op == "~=":
        # Compatible release: same major.minor, patch >= base
        return inst_parts[:2] == base_parts[:2] and inst_parts[2] >= base_parts[2]
    # Other operators — assume OK
    return True


def check_sdk_version_alignment() -> None:
    """Compare the app's declared SDK deps against installed versions."""
    pyproject = _find_app_pyproject()
    if not pyproject:
        return

    deps: list[str] = []
    project = pyproject.get("project", {})
    deps.extend(project.get("dependencies", []))
    for extras in project.get("optional-dependencies", {}).values():
        deps.extend(extras)

    installed = get_runtime_sdk_versions()
    mismatches: list[tuple[str, str, str]] = []

    for dep in deps:
        for pkg_name in SDK_PACKAGES:
            if not dep.startswith(pkg_name):
                continue
            # Strip extras like [runtime]
            spec_part = re.sub(r"\[.*?\]", "", dep)
            # Extract version specifier
            spec_match = re.search(r"([><=!~]+\s*\d+\.\d+\.\d+)", spec_part)
            if not spec_match or pkg_name not in installed:
                continue
            spec = spec_match.group(1)
            if not _satisfies_range(installed[pkg_name], spec):
                mismatches.append((pkg_name, spec, installed[pkg_name]))

    if mismatches:
        print(f"{LOG_PREFIX} SDK version mismatch detected:")
        for name, declared, inst in mismatches:
            print(f"  {name}: app declares {declared}, runtime provides {inst}")
        print(f"{LOG_PREFIX} Consider aligning your app's SDK versions with the runtime.")
